src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java

```diff
@@ -0,0 +1,271 @@
+    }
+
+    public void addInterceptor(SecurityAnalysisInterceptor interceptor) {
+        interceptors.add(Objects.requireNonNull(interceptor));
+    }
+
+    public boolean removeInterceptor(SecurityAnalysisInterceptor interceptor) {
+        return interceptors.remove(Objects.requireNonNull(interceptor));
+    }
+
+    public CompletableFuture<SecurityAnalysisReport> run(String workingVariantId, SecurityAnalysisParameters securityAnalysisParameters,
+                                                         ContingenciesProvider contingenciesProvider) {
+        Objects.requireNonNull(workingVariantId);
+        Objects.requireNonNull(securityAnalysisParameters);
+        Objects.requireNonNull(contingenciesProvider);
+        return CompletableFuture.supplyAsync(() -> {
+            String oldWorkingVariantId = network.getVariantManager().getWorkingVariantId();
+            network.getVariantManager().setWorkingVariant(workingVariantId);
+            SecurityAnalysisReport result = runSync(securityAnalysisParameters, contingenciesProvider);
+            network.getVariantManager().setWorkingVariant(oldWorkingVariantId);
+            return result;
+        });
+    }
+
+    abstract SecurityAnalysisReport runSync(SecurityAnalysisParameters securityAnalysisParameters, ContingenciesProvider contingenciesProvider);
+
+    /**
+     * Detect violations on branches and on buses
+     * @param branches branches on which the violation limits are checked
+     * @param buses buses on which the violation limits are checked
+     * @param violations list on which the violation limits encountered are added
+     */
+    protected void detectViolations(Stream<LfBranch> branches, Stream<LfBus> buses, Map<Pair<String, Branch.Side>, LimitViolation> violations) {
+        // Detect violation limits on branches
+        branches.forEach(branch -> detectBranchViolations(branch, violations));
+
+        // Detect violation limits on buses
+        buses.forEach(bus -> detectBusViolations(bus, violations));
+    }
+
+    /**
+     * Detect violation limits on one branch and add them to the given list
+     * @param branch branch of interest
+     * @param violations list on which the violation limits encountered are added
+     */
+    protected void detectBranchViolations(LfBranch branch, Map<Pair<String, Branch.Side>, LimitViolation> violations) {
+        // detect violation limits on a branch
+        // Only detect the most serious one (findFirst) : limit violations are ordered by severity
+        if (branch.getBus1() != null) {
+            branch.getLimits1(LimitType.CURRENT).stream()
+                .filter(temporaryLimit1 -> branch.getI1().eval() > temporaryLimit1.getValue())
+                .findFirst()
+                .map(temporaryLimit1 -> createLimitViolation1(branch, temporaryLimit1, LimitViolationType.CURRENT, PerUnit.SB / branch.getBus1().getNominalV(), branch.getI1().eval()))
```

Comment:
Maybe an `getUnScaleForLimitType` should be great here?

Reply:
In AbstractLfBranch API ?

Reply:
Yes? As you need with scale.